clients/drcachesim/tracer/instru.cpp

```diff
@@ -78,6 +78,72 @@ instru_t::instr_to_instr_type(instr_t *instr, bool repstr_expanded)
     return TRACE_TYPE_INSTR;
 }
 
+#ifdef AARCH64
+// XXX: this decoding logic may be moved to DR's IR for reuse.
+unsigned short
+instru_t::get_aarch64_prefetch_target_policy(ptr_int_t prfop)
+{
+    // Least significant three bits of prfop represent prefetch target and
+    // policy [... b2 b1 b0]
+    //
+    // b2 b1 : prefetch target
+    // 0b00 - L1 cache, 0b01 - L2 cache, 0b10 - L3 cache
+    //
+    // b0 : prefetch policy
+    // 0b0 - temporal prefetch, 0b1 - streaming (non-temporal) prefetch
+    return prfop & 0b111;
+}
+
+unsigned short
+instru_t::get_aarch64_prefetch_op_type(ptr_int_t prfop)
+{
+    // Bits at position 3 and 4 in prfop represent prefetch operation type
+    // [... b4 b3 ...]
+    // 0b00 - prefetch for load
+    // 0b01 - prefetch for instruction
+    // 0b10 - prefetch for store
+    return (prfop >> 3) & 0b11;
+}
+
+unsigned short
+instru_t::get_aarch64_prefetch_type(ptr_int_t prfop)
+{
+    unsigned short target_policy = get_aarch64_prefetch_target_policy(prfop);
+    unsigned short op_type = get_aarch64_prefetch_op_type(prfop);
+    switch (op_type) {
+    case 0b00: // prefetch for load
+        switch (target_policy) {
+        case 0b000: return TRACE_TYPE_PREFETCH_READ_L1;
+        case 0b001: return TRACE_TYPE_PREFETCH_READ_L1_NT;
+        case 0b010: return TRACE_TYPE_PREFETCH_READ_L2;
+        case 0b011: return TRACE_TYPE_PREFETCH_READ_L2_NT;
+        case 0b100: return TRACE_TYPE_PREFETCH_READ_L3;
+        case 0b101: return TRACE_TYPE_PREFETCH_READ_L3_NT;
+        }
+    case 0b01: // prefetch for instruction
+        switch (target_policy) {
+        case 0b000: return TRACE_TYPE_PREFETCH_INSTR_L1;
+        case 0b001: return TRACE_TYPE_PREFETCH_INSTR_L1_NT;
+        case 0b010: return TRACE_TYPE_PREFETCH_INSTR_L2;
+        case 0b011: return TRACE_TYPE_PREFETCH_INSTR_L2_NT;
+        case 0b100: return TRACE_TYPE_PREFETCH_INSTR_L3;
+        case 0b101: return TRACE_TYPE_PREFETCH_INSTR_L3_NT;
+        }
+    case 0b10: // prefetch for store
+        switch (target_policy) {
+        case 0b000: return TRACE_TYPE_PREFETCH_WRITE_L1;
+        case 0b001: return TRACE_TYPE_PREFETCH_WRITE_L1_NT;
+        case 0b010: return TRACE_TYPE_PREFETCH_WRITE_L2;
+        case 0b011: return TRACE_TYPE_PREFETCH_WRITE_L2_NT;
+        case 0b100: return TRACE_TYPE_PREFETCH_WRITE_L3;
+        case 0b101: return TRACE_TYPE_PREFETCH_WRITE_L3_NT;
+        }
+    }
+    DR_ASSERT_MSG(false, "Unsupported AArch64 prefetch operation.");
```

Comment:
[1] mentions that there may be some encodings not accessible using prfop. I'm not sure how these work. Do you suggest we simply map them to the default `TRACE_TYPE_PREFETCH`, instead of throwing an error?

[1] https://static.docs.arm.com/ddi0596/a/DDI_0596_ARM_a64_instruction_set_architecture.pdf Page 829

Reply:
> there may be some encodings not accessible using prfop

You mean 0b110 and 0b111?

Pros and cons depending on the use case.  If we did map to the default we should have a warning or debug-build-only error.  I guess I would lean toward release build mapping to generic and debug throwing an error for something like this.

Reply:
0bxx110, 0bxx111 and also 0b11xxx. 

I made changes so that the assert applies only for DEBUG builds. The offline trace test prints the error message and exits as expected. But the online test doesn't exit after printing and needs to be terminated manually. Is this expected?
`ASSERT FAILURE: /home/abhinavas/dr/src/i4328-1/clients/drcachesim/tracer/instru.cpp:144: false (Unsupported AArch64 prefetch operation.)`

Reply:
> But the online test doesn't exit after printing and needs to be terminated manually.

You mean the test app exits, but the simulator process doesn't detect the exit?  I think we try to close the pipe on an exit but there are definitely still some outstanding hangs, such as #4296.  Maybe worth a quick check to ensure the pipe is closed -- do we need a custom abort.

Reply:
I used `top` to check which process has trouble terminating.
```
 ./bin64/drrun -t drcachesim -simulator_type basic_counts -- suite/tests/bin/allasm_aarch64_prefetch
```
`allasm_aarch64_prefetch` app terminates, but `drcachesim` doesn't

But the following workaround helps. It exits as expected.
```
./bin64/drrun -t drcachesim  -dr_ops -unsafe_build_ldstex -simulator_type basic_counts -- setarch `uname -m` -R    suite/tests/bin/allasm_aarch64_prefetch
```
I confirmed that we need both `-dr_ops -unsafe_build_ldstex`  and the `setarch`

Reply:
The `-unsafe_build_ldstex` is #1698 and that would eliminate a hang in the app process, which is not what you saw: you saw a hang in the simulator process (drcachesim).

The `setarch` is strange: that is the opposite of #4296 where launching under `setarch` *caused* a hang.  On your machine does #4296 not reproduce as described in the bug?

Reply:
In this case, the hang occurs even without the `setarch`, so perhaps this is a different issue?

Another observation:
#4296 suggests running `drrun` under the `setarch`. However, in this case, the following hangs as well
```
setarch `uname -m` -R ./bin64/drrun -t drcachesim  -simulator_type basic_counts -- suite/tests/bin/allasm_aarch64_prefetch
```

Reply:
The following patch allows to reproduce this issue with the allasm_aarch64_prefetch app on debug builds.

```
clients/drcachesim/tests/allasm_aarch64_prefetch.asm
@@ -163,6 +163,7 @@ _start:
         prfum    pstl3strm, [x1]
         prfum    pstl3strm, [x1]
 
+        prfum    28, [x1]
         // Exit.
         mov      w0, #1            // stdout
         adr      x1, helloworld
```

Reply:
> On your machine does #4296 not reproduce as described in the bug?

On my machine, the issue reproduces as described. However, the workaround doesn't. All of the following hang, with the app process and `drcachesim` still showing up in `top`:
```
bin64/drrun -t drcachesim -- suite/tests/bin/simple_app
bin64/drrun -t drcachesim -- setarch `uname -m` -R suite/tests/bin/simple_app
setarch `uname -m` -R bin64/drrun -t drcachesim --  suite/tests/bin/simple_app
```

The following also hangs, but neither drcachesim nor the app show up in `top`. Weird.
```
bin64/drrun -t drcachesim  -dr_ops -unsafe_build_ldstex  -- setarch `uname -m` -R suite/tests/bin/simple_app
```

However, the following works:
```
setarch `uname -m` -R bin64/drrun -t drcachesim  -dr_ops -unsafe_build_ldstex  --  suite/tests/bin/simple_app
```

Reply:
Make sure to delete any stale pipe file before any subsequent run: that will cause a hang.  Also, when you say "in top", are you only looking at the top N sorted by cpu usage?  Be sure to look at all processes through whatever method.

Reply:
> The following also hangs, but neither drcachesim nor the app show up in top. Weird.
> 
> bin64/drrun -t drcachesim  -dr_ops -unsafe_build_ldstex  -- setarch `uname -m` -R suite/tests/bin/simple_app

`ps -x` indeed shows both `drcachesim` and the app when this command is run. Thanks.


I ran each of the commands in my last comment and verified that `/tmp/drcachesimpipe` gets deleted after I kill the previous process. Same behaviour.
There is a `/tmp/drtestpipe` that's not owned by me on the system, but that is used only by `ctest` so shouldn't be the cause here.

Reply:
This issue seems different than #4296 -- it comes up on assert failure only, and without `setarch`. Do we investigate further here or handle it in a separate issue?

Reply:
+1 for separate.

Reply:
Raised #4376 for this bug.